File: backend/app/repositories/CourseRepository.scala
package repositories

import com.datastax.oss.driver.api.core.cql.Row
import models._
import services.CassandraService
import javax.inject._
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._
import java.time.Instant
import java.util.UUID
import play.api.Logger

@Singleton
class CourseRepository @Inject()(
  cassandra: CassandraService
)(implicit ec: ExecutionContext) {
  
  private val logger = Logger(this.getClass)
  private val keyspace = "subjectplanning"
  
  def findById(id: UUID): Future[Option[Course]] = {
    val query = s"""
      SELECT * FROM $keyspace.courses 
      WHERE id = $id
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToCourse(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding course: ${e.getMessage}", e)
        None
    }
  }
  
  def findByCourseCode(courseCode: String): Future[Option[Course]] = {
    val query = s"""
      SELECT * FROM $keyspace.courses 
      WHERE course_code = '$courseCode' 
      ALLOW FILTERING
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToCourse(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding course by code: ${e.getMessage}", e)
        None
    }
  }
  
  def findAll(): Future[List[Course]] = {
    val query = s"SELECT * FROM $keyspace.courses"
    
    cassandra.executeAsync(query).map { resultSet =>
      resultSet.currentPage().asScala.map(mapRowToCourse).toList
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding all courses: ${e.getMessage}", e)
        List.empty
    }
  }
  
  def create(course: Course): Future[Course] = {
    val prerequisitesStr = course.prerequisites.map(p => s"'$p'").mkString("[", ",", "]")
    
    val query = s"""
      INSERT INTO $keyspace.courses 
      (id, course_code, course_name, credits, difficulty, prerequisites, description, created_at)
      VALUES (
        ${course.id},
        '${course.courseCode}',
        '${course.courseName.replace("'", "''")}',
        ${course.credits},
        ${course.difficulty},
        $prerequisitesStr,
        '${course.description.replace("'", "''")}',
        '${course.createdAt}'
      )
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Created course: ${course.courseCode}")
      course
    }.recover {
      case e: Exception =>
        logger.error(s"Error creating course: ${e.getMessage}", e)
        throw e
    }
  }
  
  def update(course: Course): Future[Course] = {
    val prerequisitesStr = course.prerequisites.map(p => s"'$p'").mkString("[", ",", "]")
    
    val query = s"""
      UPDATE $keyspace.courses 
      SET course_name = '${course.courseName.replace("'", "''")}',
          credits = ${course.credits},
          difficulty = ${course.difficulty},
          prerequisites = $prerequisitesStr,
          description = '${course.description.replace("'", "''")}'
      WHERE id = ${course.id}
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Updated course: ${course.id}")
      course
    }.recover {
      case e: Exception =>
        logger.error(s"Error updating course: ${e.getMessage}", e)
        throw e
    }
  }
  
  def delete(id: UUID): Future[Boolean] = {
    val query = s"""
      DELETE FROM $keyspace.courses 
      WHERE id = $id
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Deleted course: $id")
      true
    }.recover {
      case e: Exception =>
        logger.error(s"Error deleting course: ${e.getMessage}", e)
        false
    }
  }
  
  def findByIds(ids: List[UUID]): Future[List[Course]] = {
    if (ids.isEmpty) return Future.successful(List.empty)
    
    val idsStr = ids.mkString(",")
    val query = s"""
      SELECT * FROM $keyspace.courses 
      WHERE id IN ($idsStr)
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      resultSet.currentPage().asScala.map(mapRowToCourse).toList
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding courses by ids: ${e.getMessage}", e)
        List.empty
    }
  }
  
  private def mapRowToCourse(row: Row): Course = {
    val prerequisitesList = row.getList("prerequisites", classOf[String])
    
    Course(
      id = row.getUuid("id"),
      courseCode = row.getString("course_code"),
      courseName = row.getString("course_name"),
      credits = row.getInt("credits"),
      difficulty = row.getDouble("difficulty"),
      prerequisites = if (prerequisitesList != null) prerequisitesList.asScala.toList else List.empty,
      description = row.getString("description"),
      createdAt = row.getInstant("created_at")
    )
  }
}

File: backend-python/app/repositories/course_repository.py
from typing import Optional, List
from uuid import UUID
import logging
from app.models import Course
from app.services.cassandra_service import cassandra_service

logger = logging.getLogger(__name__)


class CourseRepository:
    def __init__(self):
        self.session = cassandra_service.get_session()
        self.keyspace = "subjectplanning"
    
    def find_by_id(self, course_id: UUID) -> Optional[Course]:
        """Find course by ID"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.courses 
                WHERE id = %s
            """
            result = self.session.execute(query, (course_id,))
            row = result.one()
            
            if row:
                return self._map_row_to_course(row)
            return None
        except Exception as e:
            logger.error(f"Error finding course by id: {str(e)}")
            return None
    
    def find_by_course_code(self, course_code: str) -> Optional[Course]:
        """Find course by course code"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.courses 
                WHERE course_code = %s 
                ALLOW FILTERING
            """
            result = self.session.execute(query, (course_code,))
            row = result.one()
            
            if row:
                return self._map_row_to_course(row)
            return None
        except Exception as e:
            logger.error(f"Error finding course by code: {str(e)}")
            return None
    
    def find_all(self) -> List[Course]:
        """Find all courses"""
        try:
            query = f"SELECT * FROM {self.keyspace}.courses"
            result = self.session.execute(query)
            return [self._map_row_to_course(row) for row in result]
        except Exception as e:
            logger.error(f"Error finding all courses: {str(e)}")
            return []
    
    def create(self, course: Course) -> Course:
        """Create a new course"""
        try:
            query = f"""
                INSERT INTO {self.keyspace}.courses 
                (id, course_code, course_name, credits, difficulty, prerequisites, description, created_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            self.session.execute(query, (
                course.id,
                course.course_code,
                course.course_name,
                course.credits,
                course.difficulty,
                course.prerequisites,
                course.description,
                course.created_at
            ))
            
            logger.info(f"Created course: {course.course_code}")
            return course
        except Exception as e:
            logger.error(f"Error creating course: {str(e)}")
            raise
    
    def update(self, course: Course) -> Course:
        """Update an existing course"""
        try:
            query = f"""
                UPDATE {self.keyspace}.courses 
                SET course_name = %s, credits = %s, difficulty = %s, 
                    prerequisites = %s, description = %s
                WHERE id = %s
            """
            
            self.session.execute(query, (
                course.course_name,
                course.credits,
                course.difficulty,
                course.prerequisites,
                course.description,
                course.id
            ))
            
            logger.info(f"Updated course: {course.id}")
            return course
        except Exception as e:
            logger.error(f"Error updating course: {str(e)}")
            raise
    
    def delete(self, course_id: UUID) -> bool:
        """Delete a course"""
        try:
            query = f"""
                DELETE FROM {self.keyspace}.courses 
                WHERE id = %s
            """
            self.session.execute(query, (course_id,))
            logger.info(f"Deleted course: {course_id}")
            return True
        except Exception as e:
            logger.error(f"Error deleting course: {str(e)}")
            return False
    
    def _map_row_to_course(self, row) -> Course:
        """Map Cassandra row to Course model"""
        return Course(
            id=row.id,
            course_code=row.course_code,
            course_name=row.course_name,
            credits=row.credits,
            difficulty=row.difficulty,
            prerequisites=list(row.prerequisites) if row.prerequisites else [],
            description=row.description or "",
            created_at=row.created_at
        )


# Singleton instance
course_repository = CourseRepository()
